### PR TITLE
make sourcegraph support redis password

### PR DESCRIPTION
**Pull request by alexandnpu**
Test plan: 

sourcegraph should be able to support redis with authorization.

Note:
I failed to setup my dev environment to test on my own mac, as the dev setup is not that easy enough for only the server side.
Maybe this PR is an idea to show you that sourcegraph should be able to use password to connect redis. Because in my company, every redis server should use password to connect.

### CODE REVIEW

File: pkg/redispool/redispool.go [hunk 1]
```diff
@@ -49,6 +53,50 @@ func init() {
 	}
 }
 
+func connectRedis(redisUrl string) (redis.Conn, error) {
+	parsedUrl, err := url.Parse(redisUrl); if err != nil {
+		return nil, err
+	}
+
+	var host string
+	if len(parsedUrl.Opaque) > 0 {
```
Comment (beyang) [inline pkg/redispool/redispool.go]: Can you provide an example of a URL with an opaque component that would be used here? Haven't seen one used for connecting to Redis before, and adding that example to a unit test would be beneficial.

Comment (beyang) [inline pkg/redispool/redispool.go]: Can you provide an example of a URL with an opaque component that would be used here? Haven't seen one used for connecting to Redis before, and adding that example to a unit test would be beneficial.

Comment (alexandnpu) [reply pkg/redispool/redispool.go]: Maybe `Opaque` is not a very straight forward work to understand. That part of code is handling urls whose pattern is `host:port`, such as "localhost:5000" or so.

I do not like the way `Opaque` used here, but it is in the standard library net/url/url.go

```
// A URL represents a parsed URL (technically, a URI reference).
//
// The general form represented is:
//
//	[scheme:][//[userinfo@]host][/]path[?query][#fragment]
//
// URLs that do not start with a slash after the scheme are interpreted as:
//
//	scheme:opaque[?query][#fragment]
//
// Note that the Path field is stored in decoded form: /%47%6f%2f becomes /Go/.
// A consequence is that it is impossible to tell which slashes in the Path were
// slashes in the raw URL and which were %2f. This distinction is rarely important,
// but when it is, code must not use Path directly.
// The Parse function sets both Path and RawPath in the URL it returns,
// and URL's String method uses RawPath if it is a valid encoding of Path,
// by calling the EscapedPath method.
type URL struct {
	Scheme     string
	Opaque     string    // encoded opaque data
	User       *Userinfo // username and password information
	Host       string    // host or host:port
	Path       string    // path (relative paths may omit leading slash)
	RawPath    string    // encoded path hint (see EscapedPath method)
	ForceQuery bool      // append a query ('?') even if RawQuery is empty
	RawQuery   string    // encoded query values, without '?'
	Fragment   string    // fragment for references, without '#'
}
```

any suggestions for making this part more elegant?

Comment (alexandnpu) [reply pkg/redispool/redispool.go]: I've updated my way of parsing redis connection string and added some tests. Please have a review.

**Comment by beyang** [comment]
Thanks for pushing up this PR @alexandnpu. I've incorporated this into  https://github.com/sourcegraph/sourcegraph/pull/3482, so will close this in favor of that. I expect this to ship in 3.4.

**Comment by alexandnpu** [comment]
I am really happy that you can accept my idea.

I am new to GO, and your solution is really better than mine. I did not realize that there is DialUrl interface in redigo. And the standard, https://www.iana.org/assignments/uri-schemes/prov/redis.

I learned from you. Thank you!

**Pull request closed without merging**